docs/design/generic-enhanced-purls.md

```diff
@@ -0,0 +1,233 @@
+
+# git clone command (in case file_name is absent) 
+git clone https://{repository_url}/{namespace}/{name}
+git checkout {commit_hash}
+```
+
+#### SBOM component
+
+```json
+[
+    {
+        "name": "Llama-3.1-Nemotron-70B-Instruct-HF",
+        "purl": "pkg:huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF@043235d6088ecd3dd5fb5ca3592b6913fd516027?checksum=sha256:e4ca5fad",
+        "properties": [
+            {
+                "name": "cachi2:found_by",
+                "value": "cachi2:generic"
+            }
+        ],
+        "externalReferences": [
+            {
+                "type": "distribution",
+                "url": "https://huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF"
+            }
+        ],
+        "type": "library"
+    },
+    {
+        "name": "Llama-3.1-Nemotron-70B-Instruct-HF",
+        "purl": "pkg:huggingface/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF@043235d6088ecd3dd5fb5ca3592b6913fd516027?checksum=sha256:e4ca5fad&file_name=model-00001-of-00030.safetensors",
+        "properties": [
+            {
+                "name": "cachi2:found_by",
+                "value": "cachi2:generic"
+            }
+        ],
+        "externalReferences": [
+            {
+                "type": "distribution",
+                "url": "https://huggingface.co/nvidia/Llama-3.1-Nemotron-70B-Instruct-HF/blob/fac73d3507320ec1258620423469b4b38f88df6e/model-00001-of-00030.safetensors"
+            }
+        ],
+        "type": "file"
+    }
+]
+```
+
+### OCI
+
+#### generic_artifacts.yaml
+```yaml
+metadata:
+    version: '1.0'
+artifacts:
+  - purl: pkg:oci/buildah-task@sha256%3Ab2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c?repository_url=quay.io/konflux-ci&arch=amd64&tag=latest
+    target: quakus.jar
+```
+
+#### resolving the download url
+```python
+# parse the input purl
+pkg:oci/{name}@{algorithm}%3A{digest}?repository_url={repository_url_and_namespace}[&arch={arch}][&tag={tag}]
+
+# generate the download command
+podman pull {repository_url_and_namespace}/{name}@{algorithm}:{digest}
+```
+
+#### SBOM component
+```json
+{
+    "name": "buildah-task",
+    "purl": "pkg:oci/buildah-task@sha256%3Ab2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c?repository_url=quay.io/konflux-ci&arch=amd64&tag=latest",
+    "properties": [
+        {
+            "name": "cachi2:found_by",
+            "value": "cachi2:generic"
+        }
+    ],
+    "externalReferences": [
+        {
+            "type": "distribution",
+            "url": "quay.io/konflux-ci/buildah-task@sha256:b2d6c32d1e05e91920cd4475b2761d58bb7ee11ad5dff3ecb59831c7572b4d0c"
+        }
+    ],
+    "type": "container"
+}
+```
+
+## Initial thoughts
+
+From a Cachi2 perspective, we can separate purls types into ones that are part of existing package manager ecosystems (such as nuget, composer, maven) and ones that are not (github, huggingface, oci):
```

Comment:
What benefit do we have from this separation? Do we do anything with this classification?

Reply:
We don't do anything with them. I initially had a hard time wrapping my head around non-package manager types (oci, github, huggingface), because they don't simply point to a file in a URL. I initally assumed we don't want to allow this feature to cover other existing package managers, because we'd want full support for a package manager.

I think the question here is: how flexible we want this feature to become. It started as a way to fetch a files, but do we want to extend it to OCI artifacts or even git repos?

Reply:
Any of these artifact types vary with how metadata is stored and associated with the content that is downloaded. 

Being able to specify/support purl types that could match to future package manager support is one way to simplify the required work for fetching content. The downside to this (from a users' perspective) is that the native dependency files for the package managers are not supported. If, in the future, a package manager-native way to resolve dependencies is developed, it would be an option for the Cachi2 tooling to deprecate the generic purl-based fetching mechanism. This deprecation would, of course, require a breaking change.

Reply:
I would consider flexibility to be driven by a case-by-case basis. If there is requested support for fetching some form of artifact, is it required to have the native dependency format (i.e. lockfiles)? Or is it possible/reasonable to implement the fetching with a generic purl-based approach? Implementation with a purl-based approach seems like it would at least be faster if that is an acceptable solution.

Reply:
> simplify the required work for fetching content

We want to make sure, for anything with an *existing* package manager, that we don't break builds, i.e. anything which happens *after* the fetch.

For instance: if we have to handle some arbitrary level of complexity to resolve the dependency graph we need to download, and/or some arbitrarily complex setup step is involved (e.g. a specific filesystem layout)... we're better off simply using the existing package manager ecosystem and it's lockfiles to resolve the dependency graph, get URLs, download to correct locations from those URLs, etc...

To a first approximation, cachi2 is a very smart "package manager wrapper" - our secondary brief is to make sure that e.g. `go build`, `pip install -r foo`, `yarn` (er... build? >_<) *succeed*, and our **primary** brief is that ***we*** control the downloads which the build relies on.

Reply:
I don't think that it is Cachi2's job to ensure that builds are not broken if a purl-driven package manager was used. This should be handled by users' specifications of what content to download ... whether it is by specifying a path to a lockfile or a set of purls. It does, however, seem reasonable that users expectations are set based on the choice of interface.

For example, you may have a lockfile that only has one tarball: https://github.com/stolostron/image-builder/blob/master/yarn-source/package-lock.json

A user should be able to use the npm package manager to resolve that if they want to maintain the package lock. If they don't want to maintain the package lock, they should be able to just fetch that tarball.

Similarly, if there is not support for a specific package manager in Cachi2 but a user wants to prefetch all content, they should be able to take the complexity to individually specify all resolved packages for their builds. Cachi2 wouldn't know that there is complete resolution of artifacts but that is irrelevant.

If Cachi2 wants to record how dependencies were defined (i.e. if they were resolved with a package manager or defined individually), it seems reasonable that the generated SBOM would have metadata to indicate how the specific dependency was resolved.